lib/compress/zstd_compress.c

```diff
@@ -2803,18 +2883,714 @@ static ZSTD_blockCompressor ZSTD_selectBlockCompressor(ZSTD_strategy strat, int
+            U32 offset;
+            const BYTE* match;
+            size_t forwardMatchLength = 0, backwardMatchLength = 0;
+            ldmEntry_t* bestEntry = NULL;
+            U32 tag = ZSTD_ldm_getTag(rollingHash, hBits);
+            if (tag != LDM_HASH_MASK) {
+               /* Do not insert and do not look for a match */
+               ip++;
+               continue;
+            }
+
+            /* Get the best entry and compute the match lengths */
+            {
+                ldmEntry_t* bucket =
+                    ZSTD_ldm_getBucket(hashTable,
+                                       ZSTD_ldm_getSmallHash(rollingHash,
+                                                             hBits));
+                ldmEntry_t* cur;
+                size_t bestMatchLength = 0;
+
+                for (cur = bucket; cur < bucket + LDM_BUCKET_SIZE; ++cur) {
+                    const BYTE* pMatch = cur->offset + base;
+                    if (cur->checksum != checksum ||
+                        cur->offset <= lowestIndex) {
+                        continue;
+                    } else {
+                        size_t curForwardMatchLength =
+                            ZSTD_count(ip, pMatch, iend);
+                        size_t curBackwardMatchLength, curTotalMatchLength;
+
+                        if (curForwardMatchLength < LDM_MIN_MATCH_LENGTH) {
+                            continue;
+                        }
+                        curBackwardMatchLength =
+                            ZSTD_ldm_countBackwardsMatch(
+                                    ip, anchor, pMatch, lowest);
+                        curTotalMatchLength = curForwardMatchLength +
+                                              curBackwardMatchLength;
+
+                        if (curTotalMatchLength > bestMatchLength) {
+                            bestMatchLength = curTotalMatchLength;
+                            forwardMatchLength = curForwardMatchLength;
+                            backwardMatchLength = curBackwardMatchLength;
+                            bestEntry = cur;
+                        }
+                    }
+                }
+            }
+
+            if (bestEntry == NULL) {
+                ZSTD_ldm_makeEntryAndInsertByTag(hashTable, bucketOffsets,
+                                                 rollingHash, hBits,
+                                                 current);
+                ip++;
+                continue;
+            }
+
+            matchIndex = bestEntry->offset;
+            match = base + matchIndex;
+            mLength = forwardMatchLength;
+            offset = (U32)(ip-match);
+
+            ip -= backwardMatchLength;
+            match -= backwardMatchLength;
+            mLength += backwardMatchLength;
+
+            /* Overwrite rep codes */
+            for (i = 0; i < ZSTD_REP_NUM; i++)
+                seqStorePtr->rep[i] = repToConfirm[i];
+
+            ZSTD_ldm_fillTables(cctx, anchor, base);
```

Comment:
`base` is a `const` ptr.
It does not seem correct.
You want to add new segments only,
this is adding the same segment multiple times, just enlarging it at each attempt.